File: src/main/java/com/nir/beans/PeriodicElementsGetterImpl.java
package com.nir.beans;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.nir.ui.pojos.PeriodicElement;

import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public final class PeriodicElementsGetterImpl implements PeriodicElementsGetter {
    private static List<PeriodicElement> elements = new ArrayList<>();
    private ObjectMapper objectMapper;

    public PeriodicElementsGetterImpl(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
        this.readElementsIfNoSuch();
    }

    public List<PeriodicElement> getElements() {
        this.readElementsIfNoSuch();
        return elements;
    }

    private void readElementsIfNoSuch() {
        if (elements.isEmpty()) {
            URL resource = this.getClass().getClassLoader().getResource("json/table.json");
            try {
                TypeReference<List<PeriodicElement>> type = new TypeReference<List<PeriodicElement>>() {};
                List<PeriodicElement> elements = objectMapper.readValue(resource, type);
                PeriodicElementsGetterImpl.elements.addAll(elements);
            } catch (IOException e) {
                String template = "Cant read periodic table elements: %s";
                throw new RuntimeException(String.format(template, e.getMessage()));
            }
        }
    }
}

File: src/main/kotlin/com/nir/beans/PeriodicElementsGetterImpl.kt
package com.nir.beans

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.nir.beans.PeriodicElementsGetter
import com.nir.ui.pojos.PeriodicElement
import com.nir.beans.PeriodicElementsGetterImpl
import java.io.IOException
import java.lang.RuntimeException
import java.util.ArrayList

class PeriodicElementsGetterImpl(private val objectMapper: ObjectMapper) : PeriodicElementsGetter {
    init {
        readElementsIfNoSuch()
    }

    override val elements: List<PeriodicElement>
        get() {
            readElementsIfNoSuch()
            return Companion.elements
        }

    private fun readElementsIfNoSuch() {
        if (Companion.elements.isEmpty()) {
            val resource = this.javaClass.classLoader.getResource("json/table.json")
            try {
                val type: TypeReference<List<PeriodicElement>> = object : TypeReference<List<PeriodicElement>>() {}
                val elements = objectMapper.readValue(resource, type)
                Companion.elements.addAll(elements)
            } catch (e: IOException) {
                val template = "Cant read periodic table elements: %s"
                throw RuntimeException(String.format(template, e.message))
            }
        }
    }

    companion object {
        private val elements: MutableList<PeriodicElement> = ArrayList()
    }
}